# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContext.cpp
# Commit: a505d6315623
# Full Hash: a505d6315623943fb581b7f21beece40fc6faf17
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2024-01-18 09:55:36
# Description:
#   Bug 1875193 - Don't recycle SharedSurfaces from different SurfaceFactory. r=sotaro
#   
#   It looks like we're accidentally putting SharedSurfaces that came from a different
#   SurfaceFactory back into a different one recycling. This happens because on context
#   loss in CanvasTranslator, the old GL context may go away, though the RemoteTextureOwnerClient
# ==============================================================================

diff -r 875fd9f694c9 -r a505d6315623 dom/canvas/WebGLContext.cpp
--- a/dom/canvas/WebGLContext.cpp	Thu Jan 18 05:04:12 2024 +0000
+++ b/dom/canvas/WebGLContext.cpp	Thu Jan 18 06:46:51 2024 +0000
@@ -1257,10 +1257,13 @@
 
   ownerClient->PushTexture(textureId, ownerId, keepAlive, size, surfaceFormat,
                            *desc);
-  auto recycledSurface = ownerClient->GetRecycledSharedSurface(
-      size, surfaceFormat, desc->type(), ownerId);
-  if (recycledSurface) {
-    swapChain.StoreRecycledSurface(recycledSurface);
+
+  // Look for a recycled surface that matches the swap chain.
+  while (auto recycledSurface = ownerClient->GetRecycledSharedSurface(
+             size, surfaceFormat, desc->type(), ownerId)) {
+    if (swapChain.StoreRecycledSurface(recycledSurface)) {
+      break;
+    }
   }
   return true;
 }